core/src/test/scala/tamer/utils/FakeConsumer.scala

```diff
@@ -108,18 +108,34 @@ sealed class FakeConsumer[IK, IV](
             record = new ConsumerRecord("topic", 0, offset, record.key().asInstanceOf[K], record.value().asInstanceOf[V]),
             commitHandle = (toCommit: Map[TopicPartition, Long]) =>
               (for {
-                r <- Task(new java.util.Random().nextInt(1)).map(_.longValue) // simulates the asynchronicity
-                _ <- Task(Thread.sleep(r)) // simulates the asynchronicity
                 _ <- {
                   val toCommitOffset = toCommit(partition)
                   log.info(s"consumer fakely committing offset $toCommitOffset") *>
                     inFlight.take // TODO: make sure we take the expected value
                       .tap { case (partition, value) => log.info(s"consumer fakely removing ($partition -> ${value.value()}) from in-flight") } *>
                     committed.update(_.updated(partition, Some(toCommitOffset)))
                 }
-              } yield ()).fork.unit // simulates the asynchronicity (remove this and the above when using kafka transactions)
+              } yield ()).unit
           )
         }
+        .flatMap { committableRecord =>
+          // sometimes the same message gets taken twice because, after a disconnection,
+          // the consumer loses track of the current offset and and resets to the latest
+          // committed, this effectively results in a duplicate message being processed
+          // while the same offset is committed, thus increasing the lag
+          if (r.nextFloat() < 0.005) {
+            ZStream.fromEffect(
+              inFlight.offer(
+                (
+                  new TopicPartition("topic", committableRecord.partition),
+                  new ProducerRecord("topic", committableRecord.key.asInstanceOf[IK], committableRecord.value.asInstanceOf[IV])
+                )
+              )
+            ) *>
+              ZStream(committableRecord, committableRecord.copy())
+          } else
+            ZStream(committableRecord)
+        }
```

Comment:
This would open to testing mitigations to current async bugs through unit testing, which is appealing. Corresponding functions however would have to be written as well: for example if we used `endOffsets` to decide whether to commit or not, we would have to update it because only works when starting up right now, it doesn't keep track of the fictitious offset. So right now it's only useful against code regressions.